### PR TITLE
Architecture: Fix AuthModule and Kernel Refactors

### PR DESCRIPTION
**Pull request by talekhinezh**
* Add `KernelActor` which allows modules and interepreters to access state via `system_api`
* Refactor `AuthModule` logic to use `system_api`
* Fix `AuthModule` one hop boundary logic
* Add `ScryptoInterpreter` which contains the Scrypto specific logic for executing scrypto programs
* Cleanup `SystemApi` generics

### CODE REVIEW

File: radix-engine/src/engine/actor.rs [hunk 1]
```diff
@@ -36,69 +68,25 @@ pub enum REActor {
...
+                | REActor::Function(ResolvedFunction::Native(
+                    NativeFunction::TransactionProcessor(TransactionProcessorFunction::Run)
+                ))
+        )
     }
 }
+
+/// Executing kernel level actor
+#[derive(Debug, Copy, Clone, Eq, PartialEq, TypeId, Encode, Decode)]
+pub enum KernelActor {
```
Comment (iamyulong) [inline radix-engine/src/engine/actor.rs:87]: Actor is being overloaded :)

Suggested renaming to `RestrictedMode` which further restricted what system APIs are allowed.
```rust
pub enum RestrictedMode {
   Deref,
   ScryptoFnResolver,
   AuthModule,
   /* possibly other kernel modules */
}
```

**Review by iamyulong — COMMENTED** [review]
Nice refactor overall. Only a few minor comments.

File: radix-engine/src/model/substates/auth_zone.rs [hunk 2]
```diff
@@ -1,43 +1,312 @@
...
+            HardResourceOrNonFungible::Resource(resource_address) => {
+                let proof_resource_address = proof.resource_address();
+                proof_resource_address == *resource_address
+            }
+            HardResourceOrNonFungible::SoftResourceNotFound => false,
+        }
+    }
+
+    pub fn check_auth_zones<P>(
+        mut barriers_allowed: u32,
```
Comment (iamyulong) [inline radix-engine/src/model/substates/auth_zone.rs:33]: Suggest renaming to `barrier_crossings_allowed`.

File: radix-engine/src/model/substates/auth_zone.rs [hunk 3]
```diff
@@ -1,43 +1,312 @@
...
 /// A transient resource container.
 #[derive(Debug)]
 pub struct AuthZoneSubstate {
-    pub proofs: Vec<ProofSubstate>,
-    /// IDs of buckets that act as an evidence for virtual proofs.
-    /// A virtual proof for any NonFunbigleId can be created for any ResourceAddress in the map.
-    /// Note: when a virtual proof is created,
-    /// the resources aren't actually being added to the bucket.
-    pub virtual_proofs_buckets: BTreeMap<ResourceAddress, BucketId>,
+    auth_zones: Vec<AuthZone>,
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/model/substates/auth_zone.rs:201]: good point!

File: radix-engine/src/engine/call_frame.rs [hunk 4]
```diff
@@ -178,6 +187,25 @@ impl CallFrame {
...
+        node_id: RENodeId,
+    ) -> Result<&mut HeapRootRENode, RuntimeError> {
+        self.owned_heap_nodes
+            .get_mut(&node_id)
+            .ok_or(RuntimeError::KernelError(KernelError::RENodeNotFound(
+                node_id,
+            )))
+    }
+
+    pub fn get_owned_heap_node(&self, node_id: RENodeId) -> Result<&HeapRootRENode, RuntimeError> {
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/engine/call_frame.rs:201]: yup totally agree

File: radix-engine/src/engine/call_frame.rs [hunk 5]
```diff
@@ -17,7 +17,10 @@ pub struct SubstateLock {
 pub struct CallFrame {
     /// The frame id
     pub depth: usize,
-    /// The running actor of this frame
+
+    pub kernel_actor: KernelActor,
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/engine/call_frame.rs:21]: good point

**Comment by mozi-rdx** [comment]
I was trying to elide some of the clones before, what do you think, would changes like this useful?
```diff
diff --git a/radix-engine/src/engine/call_frame.rs b/radix-engine/src/engine/call_frame.rs
index aee612cc7..a33ec1680 100644
--- a/radix-engine/src/engine/call_frame.rs
+++ b/radix-engine/src/engine/call_frame.rs
@@ -216,10 +216,7 @@ impl CallFrame {
         Ok(node_pointer)
     }
 
-    pub fn get_visible_nodes(&self) -> Vec<RENodeId> {
-        let mut node_ids: Vec<RENodeId> = self.node_refs.keys().cloned().collect();
-        let owned_ids: Vec<RENodeId> = self.owned_heap_nodes.keys().cloned().collect();
-        node_ids.extend(owned_ids);
-        node_ids
+    pub fn get_visible_nodes(&self) -> impl Iterator<Item = &RENodeId> {
+        self.node_refs.keys().chain(self.owned_heap_nodes.keys())
     }
 }
diff --git a/radix-engine/src/engine/kernel.rs b/radix-engine/src/engine/kernel.rs
index 4263ae732..a4fc6968d 100644
--- a/radix-engine/src/engine/kernel.rs
+++ b/radix-engine/src/engine/kernel.rs
@@ -896,9 +896,10 @@ where
             .map_err(RuntimeError::ModuleError)?;
         }
 
-        let node_ids = Self::current_frame_mut(&mut self.call_frames).get_visible_nodes();
-
-        Ok(node_ids)
+        Ok(Self::current_frame_mut(&mut self.call_frames)
+            .get_visible_nodes()
+            .cloned()
+            .collect())
     }
 
     fn node_drop(&mut self, node_id: RENodeId) -> Result<HeapRootRENode, RuntimeError> {
```

**Comment by talekhinezh** [comment]
> I was trying to elide some of the clones before, what do you think, would changes like this useful?
> 
> ```diff
> diff --git a/radix-engine/src/engine/call_frame.rs b/radix-engine/src/engine/call_frame.rs
> index aee612cc7..a33ec1680 100644
> --- a/radix-engine/src/engine/call_frame.rs
> +++ b/radix-engine/src/engine/call_frame.rs
> @@ -216,10 +216,7 @@ impl CallFrame {
>          Ok(node_pointer)
>      }
>  
> -    pub fn get_visible_nodes(&self) -> Vec<RENodeId> {
> -        let mut node_ids: Vec<RENodeId> = self.node_refs.keys().cloned().collect();
> -        let owned_ids: Vec<RENodeId> = self.owned_heap_nodes.keys().cloned().collect();
> -        node_ids.extend(owned_ids);
> -        node_ids
> +    pub fn get_visible_nodes(&self) -> impl Iterator<Item = &RENodeId> {
> +        self.node_refs.keys().chain(self.owned_heap_nodes.keys())
>      }
>  }
> diff --git a/radix-engine/src/engine/kernel.rs b/radix-engine/src/engine/kernel.rs
> index 4263ae732..a4fc6968d 100644
> --- a/radix-engine/src/engine/kernel.rs
> +++ b/radix-engine/src/engine/kernel.rs
> @@ -896,9 +896,10 @@ where
>              .map_err(RuntimeError::ModuleError)?;
>          }
>  
> -        let node_ids = Self::current_frame_mut(&mut self.call_frames).get_visible_nodes();
> -
> -        Ok(node_ids)
> +        Ok(Self::current_frame_mut(&mut self.call_frames)
> +            .get_visible_nodes()
> +            .cloned()
> +            .collect())
>      }
>  
>      fn node_drop(&mut self, node_id: RENodeId) -> Result<HeapRootRENode, RuntimeError> {
> ```

Yup definitely :)

**Review by iamyulong — APPROVED** [review]
LGTM. Thanks for making the changes!

**Pull request merged**